### PR TITLE
StreamDecoder: amend onXXXWriteBufferHighWatermark

### PR DESCRIPTION
**Pull request by oschaaf**
Following a report that came in via a backchannel,  remove
NOT_IMPLEMENTED_GCOVR_EXCL_LINE and add tests that will hit those.

Suspiciously, server side expectations fail with larger request body
sizes. That is pending investigation, I will file a separate issue for
that. Meantime, the test lines for that are commented out.

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by oschaaf** [comment]
Needs https://github.com/envoyproxy/nighthawk/pull/206 first, because the existing TCLAP cannot be downloaded. This will be fixed after merging in the updated Envoy dependency here.

**Comment by oschaaf** [comment]
To make sure I manually verified that with this the client will send off all the request bytes we ask it to.
The unexpectedly failing test expectations for the server side statistics occur because our test filter misbehaves when handling request data. That is tracked via https://github.com/envoyproxy/nighthawk/issues/207

File: test/integration/test_integration_basics.py [hunk 1]
```diff
@@ -342,34 +342,32 @@ def check_upload_expectations(fixture, parsed_json, expected_transmitted_bytes,
...
+    # Server side expectations start failing with larger upload sizes
+    # assertGreaterEqual(
+    #    fixture.getServerStatFromJson(server_stats,
+    #                                  "http.ingress_http.downstream_cx_rx_bytes_total"),
+    #    expected_received_bytes)
+
+  upload_bytes = 1024 * 1024 * 3
+  requests = 10
+  args = [
+      http_test_server_fixture.getTestServerRootUri(), "--duration", "100", "--rps", "100",
```
Comment (dubious90) [inline test/integration/test_integration_basics.py]: (optional) Is it possible to clean up these arguments a little for readability?  They seem to have different numbers of arguments on each line to some extent randomly, which just makes it a little bit harder to follow.

If this is just the result of straight automated tooling (e.g. automated line breaking), then don't worry about it.

File: test/integration/test_integration_basics.py [hunk 2]
```diff
@@ -342,34 +342,32 @@ def check_upload_expectations(fixture, parsed_json, expected_transmitted_bytes,
...
-  # test h1
-  parsed_json, _ = http_test_server_fixture.runNighthawkClient([
-      http_test_server_fixture.getTestServerRootUri(), "--duration", "1", "--rps", "2",
+    # Server side expectations start failing with larger upload sizes
+    # assertGreaterEqual(
+    #    fixture.getServerStatFromJson(server_stats,
+    #                                  "http.ingress_http.downstream_cx_rx_bytes_total"),
+    #    expected_received_bytes)
+
+  upload_bytes = 1024 * 1024 * 3
```
Comment (dubious90) [inline test/integration/test_integration_basics.py]: This is a pretty drastic increase in upload_bytes.  Is that what allows this test to now cover onAboveWriteBufferHighWatermark?

File: test/integration/test_integration_basics.py [hunk 3]
```diff
@@ -342,34 +342,32 @@ def check_upload_expectations(fixture, parsed_json, expected_transmitted_bytes,
...
-  # the server in between, so our expectation for received bytes on the server side is raised.
-  parsed_json, _ = http_test_server_fixture.runNighthawkClient([
-      http_test_server_fixture.getTestServerRootUri(), "--duration", "1", "--h2", "--rps", "2",
-      "--request-body-size",
-      str(upload_bytes)
-  ])
-  check_upload_expectations(http_test_server_fixture, parsed_json, upload_bytes * 2,
-                            upload_bytes * 4)
+      str(upload_bytes), "--termination-predicate",
+      "benchmark.http_2xx:%s" % str(requests), "--connections", "1", "--request-method", "POST",
```
Comment (dubious90) [inline test/integration/test_integration_basics.py]: In general, the changes to this test appear to be increasing the size of things, higher duration, rps, upload_size etc.
The description to this test from before is:
```
Test that the number of bytes we request for the request body gets reflected in the upstream
  connection transmitted bytes counter for h1 and h2.
```
Does this need an update?  I could very well be reading this wrong, but it looks like this test, while still accomplishing that, is also acting as a bit of a large load check as well.  Alternatively, would it make more sense to have this be abstracted into its own test.

File: test/integration/test_integration_basics.py (hunk 1 above)
Comment (oschaaf) [reply test/integration/test_integration_basics.py]: Yeah, so this is what fix-format/check-format makes of it.

File: test/integration/test_integration_basics.py (hunk 2 above)
Comment (oschaaf) [reply test/integration/test_integration_basics.py]: Yes indeed

File: test/integration/test_integration_basics.py (hunk 3 above)
Comment (oschaaf) [reply test/integration/test_integration_basics.py]: You are right; this deserves context: I rewrote this a bit to not terminate after x seconds, but rather after seeing 10 responses, now that we can rely on the new termination predicate feature.. This allows speedier execution and should hopefully work more stable with sluggish sanitizer-enabled runs

Comment (oschaaf) [reply test/integration/test_integration_basics.py]: (Ps the 10 seconds is an arbitrary cap to execution duration here to avoid hangs)

**Comment by oschaaf** [comment]
@dubious90 comments addressed, so this is ready for another round!

**Comment by oschaaf** [comment]
> How come we aren't just setting the watermarks to be super high (i.e. effectively unlimited)? If we don't care about buffer bounds in NH, then this makes sense. If we do care, I think we need some way to do backpressure.

I agree that being able to apply backpressure would be great to have. However, the ASSERT that happens without this needs to go away rather sooner then later :) .. shall we create an issue as a reminder for that?

**Comment by htuch** [comment]
@oschaaf please do. I'm fine with buffering for now, but let's track.

File: test/integration/test_integration_basics.py [hunk 4]
```diff
@@ -342,34 +342,32 @@ def check_upload_expectations(fixture, parsed_json, expected_transmitted_bytes,
...
-                                      "http.ingress_http.downstream_cx_rx_bytes_total"),
-        expected_received_bytes)
-
-  upload_bytes = 10000
-
-  # test h1
-  parsed_json, _ = http_test_server_fixture.runNighthawkClient([
-      http_test_server_fixture.getTestServerRootUri(), "--duration", "1", "--rps", "2",
+    # Server side expectations start failing with larger upload sizes
+    # assertGreaterEqual(
```
Comment (htuch) [inline test/integration/test_integration_basics.py]: Why are these now comments? Prefer removing if no longer relevant.

Comment (oschaaf) [reply test/integration/test_integration_basics.py]: These will get uncommented in https://github.com/envoyproxy/nighthawk/pull/208, which amends the test server and allows this to work..

**Comment by oschaaf** [comment]
> @oschaaf please do. I'm fine with buffering for now, but let's track.

Done: https://github.com/envoyproxy/nighthawk/issues/215

**Pull request merged**